src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs

```diff
@@ -723,228 +632,24 @@ private void SendHeartbeatStatusToSource(DocumentsOperationContext documentsCont
         {
-            public short TransactionMarker;
-            public ReplicationBatchItem.ReplicationItemType Type;
-
-            #region Document
-
-            public string Id;
-            public string ChangeVector;
-            public int DocumentSize;
-            public string Collection;
-            public long LastModifiedTicks;
-            public DocumentFlags Flags;
-            public BlittableJsonReaderObject Document;
-
-            #endregion
-
-            #region Counter
-
-            public long CounterValue;
-
-            public BlittableJsonReaderObject CounterValues;
-
-            #endregion
-
-            #region Attachment
-
-            public Slice Key;
-            public ByteStringContext.InternalScope KeyDispose;
-
-            public Slice Name;
-            public ByteStringContext.InternalScope NameDispose;
-
-            public Slice ContentType;
-            public ByteStringContext.InternalScope ContentTypeDispose;
-
-            public Slice Base64Hash;
-            public ByteStringContext.InternalScope Base64HashDispose;
-
-
-            #endregion
-
-            public void Dispose()
-            {
-                if (Type == ReplicationBatchItem.ReplicationItemType.Attachment)
-                {
-                    KeyDispose.Dispose();
-                    NameDispose.Dispose();
-                    ContentTypeDispose.Dispose();
-                    Base64HashDispose.Dispose();
-                }
-                else if (Type == ReplicationBatchItem.ReplicationItemType.AttachmentTombstone ||
-                         Type == ReplicationBatchItem.ReplicationItemType.RevisionTombstone)
-                {
-                    KeyDispose.Dispose();
-                }
-            }
-        }
-
-        internal struct ReplicationAttachmentStream : IDisposable
-        {
-            public Slice Base64Hash;
-            public ByteStringContext.InternalScope Base64HashDispose;
-
-            public Stream Stream;
-
-            public void Dispose()
-            {
-                Base64HashDispose.Dispose();
-                Stream.Dispose();
-            }
-        }
-
-        private unsafe ArraySegment<ReplicationItem> ReadItemsFromSource(int replicatedDocs, DocumentsOperationContext context,
-            IncomingReplicationAllocator incomingReplicationAllocator, IncomingReplicationStatsScope stats)
-        {
-            var items = ArrayPool<ReplicationItem>.Shared.Rent(replicatedDocs);
-
-            var documentRead = stats.For(ReplicationOperation.Incoming.DocumentRead, start: false);
-            var attachmentRead = stats.For(ReplicationOperation.Incoming.AttachmentRead, start: false);
-            var tombstoneRead = stats.For(ReplicationOperation.Incoming.TombstoneRead, start: false);
+            if (data.ReplicatedItems == null)
+                data.ReplicatedItems = new ReplicationBatchItem[replicatedDocs];
```

Comment:
Is this cached? Can you get a smaller value from a previous call?

Reply:
No it isn't. for now we create a new instance every time.